CRM/Civisocial/OAuthProvider.php

```diff
@@ -0,0 +1,263 @@
+<?php
+/**
+ * This class defines common functions and declares that should be overridden
+ * by each OAuth Provider
+ *
+ * To add a new OAuth Provider, add a new file named [oauth_provider_alias].php
+ * in OAuthProvider/ and extend this class. Methods declared by this class should
+ * be overriden by the new class. All letters of oauth_provider_alias should be in
+ * small case. For eg. googleplus. The class name of new OAuthProvider class should
+ * start with an uppercase letter. Rest of the letters should be in small case.
+ * For eg. Googleplus is a valid class name. GooglePlus is not a valid class name.
+ */
+
+class CRM_Civisocial_Backend_OAuthProvider {
+  /**
+   * API key/App ID/Consumer Key provided by OAuth provider
+   *
+   * @var string
+   */
+  protected $apiKey;
+
+  /**
+   * API Secret/App Secret/Consumer Secret provided by OAuth provider
+   *
+   * @var string
+   */
+  protected $apiSecret;
+
+  /**
+   * Base URL for API requests
+   *
+   * @var string
+   */
+  protected $apiUri;
+
+  /**
+   * Access Token
+   *
+   * @var mixed
+   */
+  protected $token;
+
+  /**
+   * Social user information
+   *
+   * @var array
+   */
+  protected $userProfile;
+
+  /**
+   * HTTP Status code of last API request
+   *
+   * @var string
+   */
+  protected $httpCode;
+
+  /**
+   * Default timeout
+   *
+   * @var int
+   */
+  protected $timeout = 30;
+
+  /**
+   * Default connection timeout
+   *
+   * @var int
+   */
+  protected $connectTimeout = 30;
+
+  /**
+   * Verify SSL certificate
+   *
+   * @var bool
+   */
+  protected $sslVerifyPeer = FALSE;
+
+  /**
+   * Contains the last HTTP headers returned
+   *
+   * @var string
+   */
+  protected $httpInfo = array();
+
+  /**
+   * Get social user information
+   *
+   * @return array
+   */
+  public function getUserProfile() {
+    return $this->userProfile;
+  }
+
+  /**
+   * Retrieve API credentails for the given Social Media
+   *
+   * @param string $backend
+   *      OAuth Provider short name (alias)
+   */
+  public function getApiCredentials($backend) {
+    $this->apiKey = civicrm_api3(
+            "setting",
+            "getvalue",
+            array(
+              "group" => "CiviSocial Account Credentials",
+              "name" => "{$backend}_api_key",
+            )
+        );
+    $this->apiSecret = civicrm_api3(
+            "setting",
+            "getvalue",
+            array(
+              "group" => "CiviSocial Account Credentials",
+              "name" => "{$backend}_api_secret",
+            )
+        );
+  }
+
+  /**
+   * URL to be redirected to after user authorizes
+   *
+   * @param string $backend
+   *      OAuth Provider short name (alias)
+   *
+   * @return string
+   */
+  public function getCallbackUri($backend) {
+    return rawurldecode(CRM_Utils_System::url("civicrm/civisocial/callback/{$backend}", NULL, TRUE));
```

Comment:
you can also rewrite it as 

return CRM_Utils_System::url("civicrm/civisocial/callback/{$backend}", NULL, TRUE, NULL, FALSE)